presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -256,20 +262,36 @@ private boolean appendConstraintSolvers(
     private boolean appendConstraintSolvers(
             ImmutableList.Builder<TypeConstraintSolver> resultBuilder,
             TypeSignature formalTypeSignature,
-            Type actualType,
+            TypeSignatureProvider actualTypeSignatureProvider,
             boolean allowCoercion)
     {
-        // formalTypeSignature can be categorized into one of the 4 cases below:
+        // formalTypeSignature can be categorized into one of the 5 cases below:
+        // * function type
         // * type without type parameter
         // * type parameter of type/named_type kind
         // * type with type parameter of literal/variable kind
-        // * type with type parameter of type/named_type kind
+        // * type with type parameter of type/named_type kind (except function type)
+
+        if (FunctionType.NAME.equals(formalTypeSignature.getBase())) {
+            List<TypeSignature> formalTypeParameterTypeSignatures = formalTypeSignature.getTypeParametersAsTypeSignatures();
+            resultBuilder.add(new FunctionSolver(
+                    getLambdaArgumentTypeSignatures(formalTypeSignature),
```

Comment:
I don't think "lambda" is the right name to use in `SignatureBinder`. Not all `FunctionType`s are lambdas, since I could register a `gen_function() -> function<integer, integer>` that generates functions